### PR TITLE
Initial draft of an overlord function for end-to-end analysis.

### PR DESCRIPTION
**Pull request by LTLA**
Also performs parallelization of the t-SNE, UMAP and clustering steps.

### CODE REVIEW

**Comment by LTLA** [comment]
The big question here is what we want this interface to look like. Some organization is necessary to avoid a proliferation of `namespace_*`-like argument names; same for the results.

I have some ideas about organizing the arguments more hierarchically, e.g., for `scranpy.analyze()` to call `scranpy.quality_control.analyze()` which handles the coordination of the QC-related bits.

**Comment by jkanche** [comment]
couple of thoughts, 

- we can define arguments as `namedtuple` for each step, then the analyze function for each step takes `**args` and we can let python do the unpacking. 
- may be the result can be a `SingleCellExperiment` object?, few results have an obvious place, few don't.

File: src/scranpy/argtypes.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    size_factors: Optional[np.ndarray] = None
+    center: bool = True
+    allow_zeros: bool = False
+    allow_non_finite: bool = False
+
+
+@dataclass(slots=True)
+class RnaQcArgs(BaseArgs):
+    subsets: Optional[Mapping] = None
+    block: Optional[Sequence] = None
```
Comment (jkanche) [inline src/scranpy/argtypes.py:110]: probably global arguments

and these args shouldn't extend base args

**Comment by jkanche** [comment]
lets wait on this until I finish #27

File: src/scranpy/analyze.py [hunk 2]
```diff
@@ -0,0 +1,563 @@
...
+    return {"x": res.x, "y": res.y}
+
+
+def run_umap(obj, opts):
+    res = dimred.run_umap(obj, opts)
+    return {"x": res.x, "y": res.y}
+
+
+@dataclass
+class QualityControlOpts:
```
Comment (LTLA) [inline src/scranpy/analyze.py:36]: Why are these at the top level, and not inside each subdirectory's `analyze.py`?

File: src/scranpy/analyze.py [hunk 3]
```diff
@@ -0,0 +1,563 @@
...
+__copyright__ = "ltla"
+__license__ = "MIT"
+
+
+def run_tsne(obj, opts):
+    res = dimred.run_tsne(obj, opts)
+    return {"x": res.x, "y": res.y}
+
+
+def run_umap(obj, opts):
```
Comment (LTLA) [inline src/scranpy/analyze.py:30]: Why this function exists?

File: src/scranpy/analyze.py [hunk 4]
```diff
@@ -0,0 +1,563 @@
...
+    center: bool = norm.LogNormalizeCountsArgs.center
+    allow_zeros: bool = norm.LogNormalizeCountsArgs.allow_zeros
+    allow_non_finite: bool = norm.LogNormalizeCountsArgs.allow_non_finite
+
+
+@dataclass
+class FeatureSelectionOpts:
+    """Feature selection.
+
+    Attributes:
```
Comment (LTLA) [inline src/scranpy/analyze.py:81]: Why document it here? Just refer people down to the individual `Args`.

File: src/scranpy/analyze.py [hunk 5]
```diff
@@ -0,0 +1,563 @@
...
+        span (float, optional): Span to use for the LOWESS trend fitting.
+            Defaults to 0.3.
+        num_hvgs (int, optional): Number of HVGs to pick. Defaults to 4000.
+    """
+
+    span: float = feat.ModelGeneVariancesArgs.span
```
Comment (LTLA) [inline src/scranpy/analyze.py:87]: Why are these called `Args` and the aggregates called `Opts`?

File: src/scranpy/analyze.py [hunk 6]
```diff
@@ -0,0 +1,563 @@
...
+        num_epochs (int, optional): Number of epochs to run. Defaults to 500.
+    """
+
+    min_dist: float = dimred.InitializeUmapArgs.min_dist
+    num_neighbors: int = dimred.InitializeUmapArgs.num_neighbors
+    num_epochs: int = dimred.InitializeUmapArgs.num_epochs
+
+
+@dataclass
+class SharedNearestNeighborOpts:
```
Comment (LTLA) [inline src/scranpy/analyze.py:157]: Shouldn't this be `ClusteringOpts`?

File: src/scranpy/analyze.py (hunk 3 above)
Comment (jkanche) [reply src/scranpy/analyze.py]: these are gone

**Comment by jkanche** [comment]
@LTLA

**Comment by LTLA** [comment]
options lgtm. what about the outputs? we should use a similar composibility concept for the initial `analyze` output, and then a separate function that massages it lossily into a SCE, e.g., **scran.chan**'s [`marshalToSCE`](https://github.com/LTLA/scran.chan/blob/master/R/marshalToSCE.R).

**Pull request merged**